### ISSUE TITLE
Failed to delete temp directory

### ISSUE THREAD

**Issue by kuanox** [spring-io/initializr#862]
Hello again,

A few days ago, report a problem with the tests, now I want to report a problem with temporary files when performing the tests.

```
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in io.spring.initializr.generator.language.AnnotationTests
[INFO] Running io.spring.initializr.generator.language.groovy.GroovySourceCodeWriterTests
mar 12, 2019 2:51:49 PM org.junit.jupiter.engine.execution.JupiterEngineExecutionContext close
GRAVE: Caught exception while closing extension context: org.junit.jupiter.engine.descriptor.MethodExtensionContext@5dda14d0
java.io.IOException: Failed to delete temp directory C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744. The following paths could not be deleted (see suppressed exceptions for details): , project-7340008798081558163, project-7340008798081558163\com, project-7340008798081558163\com\example, project-7340008798081558163\com\example\Test.groovy
        at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.createIOExceptionWithAttachedFailures(TempDirectory.java:206)
        at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:167)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.execution.ExtensionValuesStore.closeAllStoredCloseableValues(ExtensionValuesStore.java:61)
        at org.junit.jupiter.engine.descriptor.AbstractExtensionContext.close(AbstractExtensionContext.java:73)
        at org.junit.jupiter.engine.execution.JupiterEngineExecutionContext.close(JupiterEngineExecutionContext.java:53)
        at org.junit.jupiter.engine.descriptor.JupiterTestDescriptor.cleanUp(JupiterTestDescriptor.java:191)
        at org.junit.jupiter.engine.descriptor.JupiterTestDescriptor.cleanUp(JupiterTestDescriptor.java:54)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$cleanUp$9(NodeTestTask.java:151)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.cleanUp(NodeTestTask.java:151)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:83)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
        at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
        at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
        Suppressed: java.nio.file.DirectoryNotEmptyException: C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744
                at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
                at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
                at java.nio.file.Files.delete(Files.java:1126)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.deleteAndContinue(TempDirectory.java:187)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:182)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:173)
                at java.nio.file.Files.walkFileTree(Files.java:2688)
                at java.nio.file.Files.walkFileTree(Files.java:2742)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:173)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:165)
                ... 44 more
        Suppressed: java.nio.file.DirectoryNotEmptyException: C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744\project-7340008798081558163
                at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
                at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
                at java.nio.file.Files.delete(Files.java:1126)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.deleteAndContinue(TempDirectory.java:187)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:182)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:173)
                at java.nio.file.Files.walkFileTree(Files.java:2688)
                at java.nio.file.Files.walkFileTree(Files.java:2742)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:173)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:165)
                ... 44 more
        Suppressed: java.nio.file.DirectoryNotEmptyException: C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744\project-7340008798081558163\com
                at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
                at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
                at java.nio.file.Files.delete(Files.java:1126)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.deleteAndContinue(TempDirectory.java:187)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:182)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:173)
                at java.nio.file.Files.walkFileTree(Files.java:2688)
                at java.nio.file.Files.walkFileTree(Files.java:2742)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:173)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:165)
                ... 44 more
        Suppressed: java.nio.file.DirectoryNotEmptyException: C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744\project-7340008798081558163\com\example
                at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
                at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
                at java.nio.file.Files.delete(Files.java:1126)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.deleteAndContinue(TempDirectory.java:187)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:182)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath$1.postVisitDirectory(TempDirectory.java:173)
                at java.nio.file.Files.walkFileTree(Files.java:2688)
                at java.nio.file.Files.walkFileTree(Files.java:2742)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.deleteAllFilesAndDirectories(TempDirectory.java:173)
                at org.junit.jupiter.engine.extension.TempDirectory$CloseablePath.close(TempDirectory.java:165)
                ... 44 more
        Suppressed: java.nio.file.FileSystemException: C:\Users\NB-BAC~1\AppData\Local\Temp\junit6802267678509260744\project-7340008798081558163\com\example\Test.groovy: El proceso no tiene acceso al archivo porque est▒ siendo utilizado por otro proceso.
```

I use Windows10, Java 8, IntelliJ

Regards

**Comment by DawN1ng** [comment]
Have the same issue
Win10 Java8

**Comment by snicoll** [comment]
@sbrannen this looks like a potential JUnit bug to me. What do you think?

**Comment by wilkinsona** [comment]
It could also be our bug. Unlike Linux-like platforms, Windows will not delete a file with open handles. Perhaps we haven't closed a writer properly?

**Comment by sbrannen** [comment]
I'm not sure if this is a bug in JUnit per se, but I do believe that JUnit could perhaps be more lenient in such cases.

I've heard of similar issues with the `@TempDir` support on Windows, and I think it might be an issue with permissions. 

@sormuras, weren't you sorting out a similar issue on Windows?

@marcphilipp, what do you think about switching to a "best effort" cleanup approach and simply logging things that could not be deleted instead of failing the test?

Perhaps the upcoming JUnit Jupiter 5.4.1 could improve this.

**Comment by marcphilipp** [comment]
I think it should be strict by default in order to catch subtle bugs in the code under test, e.g. an unclosed stream. But changing to a lenient mode via the annotation (sth. like `@TempDir(cleanupMode = LENIENT)`) or a configuration parameter is certainly sth. we can consider for 5.4.1.

@sbrannen Could you please open an issue in the junit5 repo for this?

**Comment by sormuras** [comment]
> It could also be our bug. Unlike Linux-like platforms, Windows will not delete a file with open handles. Perhaps we haven't closed a writer properly?

This is true for 99% of my cases. Especially, I almost always forget to wrap the "new" stream-based helpers offered by `Files`  into a try-with-resource block.

> @sormuras, weren't you sorting out a similar issue on Windows?

Yes. But all of them were due to misbehaving test code. Or insufficient access rights ... contrived by and for the test case.

Thus, I tend to agree with @marcphilipp here: being strict and fail such tests is a good default behaviour.

**Comment by sormuras** [comment]
> [... another process has `Test.groovy` open [...]

Some (Gradle) build cache / daemon involved?

**Comment by sbrannen** [comment]
> @sbrannen Could you please open an issue in the junit5 repo for this?

See https://github.com/junit-team/junit5/issues/1814

**Comment by sbrannen** [comment]
FWIW, it looks like the following might be the cause:

https://github.com/spring-io/initializr/blob/9f4c6a53261782a5187cbe18254f85c41c7c63c1/initializr-generator/src/test/java/io/spring/initializr/generator/language/groovy/GroovySourceCodeWriterTests.java#L231-L237

According to the Javadoc for `org.springframework.util.StreamUtils.copyToString(InputStream, Charset)`:

> Copy the contents of the given InputStream into a String. Leaves the stream open when done.

... the new `FileInputStream` is never closed.

**Comment by marcphilipp** [comment]
So there was a bug waiting to be discovered after all. 😊

**Comment by snicoll** [comment]
There is a first batch of change in ec6ac4b but we'll setup CI on Windows to make sure this is properly addressed (see #865)

**Comment by snicoll** [comment]
This has been fixed as part of #879
